tgfloofbot/plugins/administration/commands.py

```diff
@@ -0,0 +1,117 @@
+from telegram.ext import CallbackContext, CallbackQueryHandler
+
+from ... import loader
+
+from ...client import TGFloofbotClient
+from ...helpers import em
+from ...logger import LOG
+
+from . import models
+from . import exceptions
+
+@loader.custom
+def administration_custom(client: TGFloofbotClient):
+
+    admin_group_id = client.config.main_group
+
+    def _resolve_admin_group_id(update):
+        try:
+            admin_group = client.updater.bot.get_chat(chat_id=admin_group_id)
+        except telegram.error.TelegramError:
+            LOG.critical("Unable to resolve the group ID set in the config! Defaulting to context group ID.")
+            admin_group = update.effective_chat
+
+        return admin_group
+
+    class WarnCommandArgs(pydantic.BaseModel):
+        bad_user: Optional[str] = pydantic.Field(
+            None, description="A raw user ID")
+        warn_message: Optional[str] = pydantic.Field(
+            None, description="The warning reason")
+
+    class UsernoteCommandArgs(pydantic.BaseModel):
+        bad_user: Optional[str] = pydantic.Field(
+            None, description="A raw user ID")
+        warn_message: Optional[str] = pydantic.Field(
+            None, description="The note")
+
+    @loader.command(name="warn", help="warn a user", admin=True)
+    def warn_command(
+        client: TGFloofbotClient,
+        update: Update,
+        context: CallbackContext,
+        args: WarnCommandArgs
+    ) -> None:
+        warn_helper(client, update, context, args, False)
+        return
+
+    @loader.command(name="note", help="adds a moderation note for a user", admin=True)
+    def usernote_command(
+        client: TGFloofbotClient,
+        update: Update,
+        context: CallbackContext,
+        args: UsernoteCommandArgs
+    ) -> None:
+        warn_helper(client, update, context, args, True)
+        return
+
+    def warn_helper(client, update, context, args, is_note):
+        user = update.effective_user
+        chat = update.effective_chat
+        admin_chat = _resolve_admin_group_id(Update)
+
+        if admin_chat.type != "group":  # group only command
+            context.bot.send_message(chat_id=chat, text="This command can only be used for groups.")
+            return
+        
+        if args.bad_user:
+            try:
+                bad_user = client.updater.bot.get_chat_member(chat_id=admin_chat.id, user_id=args.bad_user).user  # type: ignore
+                if args.warn_message:
+                    reason = args.warn_message
+                else:
+                    reason = "No reason provided."
+            except telegram.error.BadRequest:
+                raise exceptions.UserNotFoundException(args.bad_user)
+                return
+        else: 
+            context.bot.send_message(chat_id=chat.id, text="Invalid arguments. Please provide a user ID/mention and a reason.")
+            return
+
+        try:
```

Comment:
I'd argue that we shouldn't wrap this in a try/except block. We can extend future functionality of the core to message bot system administrators when any unhandled exception is raised, as this would be a pretty severe issue.

Reply:
removed